[Doc] add memory suggestion (#454)
logstash-plugins/logstash-input-beats: docs/index.asciidoc

endif::[]

[id="plugins-{type}s-{plugin}-memory"]
===== Memory usage

> This plugin uses direct memory which is not included in the heap memory.

Reviewer: ```suggestion
This plugin uses "off-heap" direct memory _in addition to_ heap memory.
```

> By default, it is the same as the heap size (Xmx), for example, setting 10GB Xmx means setting 10GB of direct memory.

Reviewer: ```suggestion
By default, a JVM's off-heap direct memory limit is the same as the heap size.
For example, setting a heap size of 10GB without setting the direct memory limit will result in the direct limit _also_ being 10GB for a total of 20GB.
```

> This plugin uses direct memory which is not included in the heap memory.
> By default, it is the same as the heap size (Xmx), for example, setting 10GB Xmx means setting 10GB of direct memory.

Reviewer 2: ```suggestion
This plugin uses direct memory which is allocated in addition to that allocated to the heap.

By default, this is set to the same value as allocated to the heap via the `Xmx` flags. For example, setting `-Xmx10G` will allocate `10GB` for heap and an additional `10GB` for direct memory, for a total of `20GB` allocated.  
```

> Make sure heap and direct memory does not exceed the total size of the server. You can set the amount of direct memory with -XX:MaxDirectMemorySize.

Reviewer: ```suggestion
Make sure heap and direct memory combined does not exceed the total memory available on the server.
You can set the amount of direct memory with `-XX:MaxDirectMemorySize` in {logstash-ref}/jvm-settings.html[Logstash's JVM Settings].
```

We should also add an off-heap section to [Logstash's JVM Settings](https://github.com/elastic/logstash/blob/main/docs/static/config-details.asciidoc).

> This plugin uses direct memory which is not included in the heap memory.
> By default, it is the same as the heap size (Xmx), for example, setting 10GB Xmx means setting 10GB of direct memory.
> Make sure heap and direct memory does not exceed the total size of the server. You can set the amount of direct memory with -XX:MaxDirectMemorySize.

Reviewer 3: ```suggestion
This plugin uses direct memory (bypassing the CPU) to improve performance. 
Direct memory and heap size (Xmx) allocations are separate, but are the same by default. 
For example, setting 10GB Xmx sets direct memory to 10GB.
You can change the amount of direct memory with -XX:MaxDirectMemorySize.

NOTE: Be sure that heap and direct memory total allocation does not exceed the total size of the server.
```

> Make sure heap and direct memory does not exceed the total size of the server. You can set the amount of direct memory with -XX:MaxDirectMemorySize.

Reviewer 2: Maybe add a link to the [jvm settings](https://www.elastic.co/guide/en/logstash/current/jvm-settings.html) doc in logstash core?

We should also update that document too, along the lines of this [issue](https://github.com/elastic/logstash/issues/12622)

Author: Adding a link sounds good. Users mostly come across direct memory with beats-input. I will update the JVM settings page. Thanks for the pointer

endif::[]

[id="plugins-{type}s-{plugin}-memory"]
===== Memory usage

This plugin uses "off-heap" direct memory in addition to heap memory.
By default, a JVM's off-heap direct memory limit is the same as the heap size.
For example, setting `-Xmx10G` without setting the direct memory limit will allocate `10GB` for heap and an additional `10GB` for direct memory, for a total of `20GB` allocated. 

> You can set the amount of direct memory with `-XX:MaxDirectMemorySize` in {logstash-ref}/jvm-settings.html[Logstash's JVM Settings]. Consider setting it to half of the heap size. Setting too low direct memory decreases the performance of ingestion.

Reviewer 3: ```suggestion
You can set the amount of direct memory with `-XX:MaxDirectMemorySize` in {logstash-ref}/jvm-settings.html[Logstash JVM Settings]. 
Consider setting direct memory to half of the heap size. 
Setting direct memory too low decreases the performance of ingestion.
```

> NOTE: Be sure that heap and direct memory combined does not exceed the total memory available on the server to avoid OutOfDirectMemoryError

Reviewer 3: What about "... to avoid an OutOfDirectMemoryError." (singular) or "... to avoid errors."  (plural)
